pkg/derivation/kv.go

```diff
@@ -0,0 +1,94 @@
+package derivation
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"os"
+)
+
+type KVStore interface {
+	Get(key string) ([]byte, error)
+	Set(key string, value []byte) error
+}
+
+// File system KV
+
+type FSKVStore struct{}
+
+func (kv *FSKVStore) Get(key string) ([]byte, error) {
+	return os.ReadFile(key)
+}
+
+func (kv *FSKVStore) Set(key string, value []byte) error {
+	return fmt.Errorf("writing not supported")
+}
```

Comment:
I'd probably start with a simple hashmap.

Reading from random places in the filesystem isn't good style, especially when used in testcases.

Reply:
Also, this will give us the opportunity to think about how things are added into the KV store. Is it ReadDerivation that'd get a reference passed in?

Reply:
I needed the file system implementation to try this out on my local Nix store.

Reply:
We might want to limit it to read from `nixpath.StoreDir` or some passed in prefix?

Reply:
Yes, for `FSKVStore`. But I think we need a different structure:

I can see some sort of component (let's call it `DrvStore` for now) that stores 2 things in a K/V structure:

 - a map from .drv path to .drv file contents
 - a map from .drv path to (intermediate) hashes.

The FSKVStore would be one possible candidate for the first concern, if you want to read .drv files on the local machine (and restricting that to `nixpath.StoreDir`, and only legal `.drv` paths definitely makes sense).

However, another valid usecase is parsing `.drv` files not on disk, but parsing .drv files from a binary cache for example.
For this, we want to have an alternative implementation, that might just be living in memory.

The second KV holds intermediate hashes. It is what's queried frequently "from the outside", and internally for hash calculation. We probably need an interface to insert a-terms into a `DrvStore` too (in the case of memory K/V, we certainly do), so there we can precompute intermediate hashes on insert.

In the case of a `DrvStore` backed the local filesystem (`FSKVStore`), we probably don't want to fail on inserting (because it's Nix that owns these files), but we should still populate a cache of intermediate hashes, whenever there's a query about a certain .drv.